### PR TITLE
Fix some unjustified disputes

### PR DESCRIPTION
**Pull request by eskimor**
Fixes https://github.com/paritytech/polkadot/issues/6057

- [ ] Fix dispute test (does it use a dispute reason, that no longer is a reason for dispute?)
- [ ] Create follow-up ticket for cleanup.

### CODE REVIEW

File: node/core/dispute-coordinator/src/initialized.rs [hunk 1]
```diff
@@ -198,59 +198,62 @@ impl Initialized {
 			gum::trace!(target: LOG_TARGET, "Waiting for message");
 			let mut overlay_db = OverlayedBackend::new(backend);
 			let default_confirm = Box::new(|| Ok(()));
-			let confirm_write =
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/initialized.rs:201]: Some indentation was messed up here. cargo fmt does not seem to fix that?

File: node/primitives/src/lib.rs [hunk 2]
```diff
@@ -219,8 +219,6 @@ pub enum InvalidCandidate {
 	ParamsTooLarge(u64),
 	/// Code size is over the limit.
 	CodeTooLarge(u64),
-	/// Code does not decompress correctly.
```
Comment (tdimitrov) [inline node/primitives/src/lib.rs:222]: `CodeDecompressionFailure` is referenced on a few more places (this causes the build errors).

**Review by tdimitrov — APPROVED** [review]

File: node/core/candidate-validation/src/lib.rs [hunk 3]
```diff
@@ -570,8 +570,7 @@ async fn validate_candidate_exhaustive(
 				"ambiguous worker death".to_string(),
 			))),
 		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
-
+			Err(ValidationFailed(e)),
```
Comment (pepyakin) [inline node/core/candidate-validation/src/lib.rs:573]: Let's leave a comment here.

File: node/core/candidate-validation/src/lib.rs [hunk 4]
```diff
@@ -527,8 +527,9 @@ async fn validate_candidate_exhaustive(
 		Err(e) => {
 			gum::info!(target: LOG_TARGET, ?para_id, err=?e, "Invalid candidate (validation code)");
 
-			// If the validation code is invalid, the candidate certainly is.
-			return Ok(ValidationResult::Invalid(InvalidCandidate::CodeDecompressionFailure))
+			// Code already passed pre-checking, if decompression fails now this most likley means
+			// some local corruption happened.
+			return Err(ValidationFailed("Code decompression failed".to_string()))
```
Comment (pepyakin) [inline node/core/candidate-validation/src/lib.rs:532]: Does every network now have pre-checking turned on?

Comment (ordian) [reply node/core/candidate-validation/src/lib.rs:532]: On the node side, as soon as people upgrade to 0.9.30 (#5977). In the runtime, Kusama and Polkadot still don't.
>  pvfCheckingEnabled: false

Someone needs to submit a motion I suppose.

**Comment by eskimor** [comment]
If we merge without pre-checking enabled, then in case of a broken parachain upgrade we would run into liveness issues as disputes won't get resolved. This is unfortunate as delaying the merge will delay enabling of slashing :grimacing: @slumber is it correct that only the runtime needs to be enacted for pre-checking to be enabled on both Kusama and Polkadot?

**Comment by slumber** [comment]
> is it correct that only the new runtime needs to be enacted for pre-checking to be enabled on both Kusama and Polkadot?

If prechecking is enabled without https://github.com/paritytech/polkadot/pull/5977 then paras upgrades will be broken, no one would be able to submit votes

If it goes into the release however, enabling prechecking on Kusama & Polkadot is a matter of changing a configuration value (+ session change)
Note that we wanted to break in prechecking on Rococo for some time before enabling it on non-testnet chains

File: node/core/candidate-validation/src/lib.rs [hunk 5]
```diff
@@ -570,8 +570,11 @@ async fn validate_candidate_exhaustive(
 				"ambiguous worker death".to_string(),
 			))),
 		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
-
+		// In principle if preparation of the `WASM` fails, the current candidate can not be the
+		// reason for that. So we can't say whether it is invalid or not in addition with
+		// pre-checking enabled only valid runtimes should ever get enacted, so we can be
+		// reasonably sure that this is some local problem on the current node.
```
Comment (slumber) [inline node/core/candidate-validation/src/lib.rs:576]: Indentation is off

**Comment by eskimor** [comment]
> Note that we wanted to break in prechecking on Rococo for some time before enabling it on non-testnet chains

Yep, but it is already running there for a while already - right? Let's get this to production asap, otherwise this pushes back on slashing too much. 0.9.31 runtime should have it enabled already.

File: node/core/candidate-validation/src/lib.rs (hunk 5 above)
Comment (slumber) [reply node/core/candidate-validation/src/lib.rs:576]: wrapping in `{}` block should help

**Review by slumber — APPROVED** [review]

**Comment by eskimor** [comment]
Depends on #6212

File: node/core/candidate-validation/src/lib.rs [hunk 6]
```diff
@@ -569,9 +569,13 @@ async fn validate_candidate_exhaustive(
 			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(
 				"ambiguous worker death".to_string(),
 			))),
-		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) =>
-			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(e))),
-
+		Err(ValidationError::InvalidCandidate(WasmInvalidCandidate::PrepareError(e))) => {
```
Comment (eskimor) [inline node/core/candidate-validation/src/lib.rs:572]: Actually now that I revisit this, this should be fixed at one layer down already: `WasmInvalidCandidate` is already the wrong error. The candidate is not at fault if preparation failed.

**Review by m-cat — APPROVED** [review]

File: node/core/candidate-validation/src/lib.rs (hunk 6 above)
Comment (eskimor) [reply node/core/candidate-validation/src/lib.rs:572]: If this is merged, actually more cleanup will be needed: 

[`PrecheckOutcome::Failed`](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/subsystem-types/src/messages.rs#L97) and the related [`PrepareError`](https://github.com/paritytech/polkadot/blob/be487ae3f8b3fda3cb0a2efe27b1500b550d1ae4/node/core/pvf/src/error.rs#L133).

Reason we no longer differentiate. During pre-checking we always assume invalid, just to be on the safe side. During preparation, we never want to raise a dispute if preparation fails (as it previously passed pre-checking), so we never want it to be of state `InvalidCandidate`.

Comment (eskimor) [reply node/core/candidate-validation/src/lib.rs:572]: Differentiation between the two error types (deterministic and non deterministic) still makes sense. For deterministic errors happening in preparation, we should print a big fat error.

Comment (mrcnski) [reply node/core/candidate-validation/src/lib.rs:572]: > Differentiation between the two error types (deterministic and non deterministic) still makes sense. For deterministic errors happening in preparation, we should print a big fat error.

You can call `is_deterministic()` on `PrepareError`. (Semantically it made sense to me to have separate error types, but others didn't think it was necessary.)

Comment (eskimor) [reply node/core/candidate-validation/src/lib.rs:572]: Yes. I arrived at the conclusion that the distinction is very useful. Non spurious errors, suggest some major issue like a hardware fault - the very least, they should have a higher log level.

**Review by ordian — APPROVED** [review]

**Comment by ordian** [comment]
Let's merge this?

**Comment by eskimor** [comment]
Polkadot does not yet have [prechecking enabled](https://polkadot.polkassembly.io/motion/324).

**Pull request merged**